index.bs

```diff
@@ -1,164 +1,194 @@
+*TODO: should we allow pointing to an external DWARF file that does not have embedded DWARF, but instead points to another external DWARF file? If so, we should mention something about allowing consumer implementations to enforce their own arbitrary max depth, and also say something about cycles.*
 
-It is also trivial to map WebAssembly bytecode offset to the derived native PC, e.g. during AOT or JIT compilation, for a use in a native debugger. 
+*TODO: should we allow heterogeneous embedded and external DWARF? Eg, have debug_line embedded, but debug_info external?*
 
-### .debug_line section ### {#debug_line}
+# Consuming and Generating DWARF for WebAssembly Code # {#webassembly-code}
 
-The `.debug_line` section allows mapping between bytecode offset and original source. The relationship between these location types is many-to-one: multiple bytecode offsets may correspond a single original source location [[wasm-debugging-capabilities/#enumerate-location-mappings|ᴿenumerate-location-mappings]].
+Note: Some DWARF constructs don't map one-to-one onto WebAssembly constructs. We strive to enumerate and resolve any ambiguities here.
 
-This information is complete to produce wasm source map file [[!source-maps]].
+## Code Addresses ## {#pc}
 
-### .debug_info section code ranges ### {#debug_info_code}
+Note: DWARF associates various bits of debug info with particular locations in the program via its code address (instruction pointer or PC). However, WebAssembly's linear memory address space does not contain WebAssembly instructions.
 
-The `.debug_info` section contains information about a compile unit, that is comprised of subprogram/functions entries. Individual function entry contains information about its parameters and variables. This information includes their types [[wasm-debugging-capabilities/#type-of-binding|ᴿtype-of-binding]], the bindings in scopes at a certain locations [[wasm-debugging-capabilities/#bindings-within-a-scope|ᴿbindings-within-a-scope]] and how to calculate a variable value using DWARF expression [[wasm-debugging-capabilities/#reconstruct-a-bindings-value|ᴿreconstruct-a-bindings-value]].
+Wherever a code address (see 2.17 of [[DWARF]]) is used in DWARF for WebAssembly, it must be the offset of an instruction relative within the `Code` section of the WebAssembly file. The DWARF is considered malformed if a PC offset is between instruction boundaries within the `Code` section.
 
-The inlined function is expressed inside the `.debug_info` function entries [[wasm-debugging-capabilities/#inlined-functions|ᴿinlined-functions]] and also contain a proper mapping in the `.debug_line` section.
+*TODO: have we considered using i th instruction instead of offsets? This would result in smaller numbers that would encode as smaller LEB128s, and might be easier for consumers and generators. If we did consider that and don't want it, it would be good to have a note here describing why we didn't choose that alternative design.*
```

Comment:
I was thinking i<sup>th</sup> instruction in the whole code section, rather than i<sup>th</sup> instruction in each function body.

I think it is worth considering that alternative design, but what I think is even more important is documenting alternatives that have been considered and why they were not ultimately chosen in informative "Note:" blocks in this document. This is the goal of a lot of my TODOs and questions: not actually advocating for the alternatives, but advocating for more documentation of *why* X was chosen instead of Y.